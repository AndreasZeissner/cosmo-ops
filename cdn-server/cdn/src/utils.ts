export async function streamToJSON(readableStream: ReadableStream) {
  const reader = readableStream.getReader();
  const decoder = new TextDecoder();
  let result = '';

  while (true) {
    const { done, value } = await reader.read();

    if (done) {
      break;
    }

    result += decoder.decode(value, { stream: true });
  }

  // Parse the accumulated result string into JSON
  return JSON.parse(result);
}
