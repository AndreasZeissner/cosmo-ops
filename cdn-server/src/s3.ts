import { GetObjectCommand, HeadObjectCommand, NoSuchKey, NotFound, S3Client } from '@aws-sdk/client-s3';
import { BlobNotFoundError, BlobObject, BlobStorage } from '@wundergraph/cosmo-cdn';
import { Context } from 'hono';

/**
 * Retrieves objects from S3 given an S3Client and a bucket name
 */
class S3BlobStorage implements BlobStorage {
  constructor(
    private s3Client: S3Client,
    private bucketName: string,
  ) {}

  async getObject({
    context,
    key,
    cacheControl,
  }: {
    context: Context;
    key: string;
    cacheControl?: string;
  }): Promise<BlobObject> {
    const command = new GetObjectCommand({
      Bucket: this.bucketName,
      Key: key,
      ResponseCacheControl: cacheControl,
    });

    try {
      const resp = await this.s3Client.send(command);
      if (resp.$metadata.httpStatusCode !== 200) {
        throw new Error(`Failed to retrieve object from S3: ${resp}`);
      }

      if (!resp.Body) {
        throw new Error(`Failed to retrieve object from S3: ${resp}`);
      }

      return {
        metadata: resp.Metadata,
        stream: resp.Body.transformToWebStream(),
      };
    } catch (e: any) {
      if (e instanceof NoSuchKey) {
        throw new BlobNotFoundError(`Failed to retrieve object from S3: ${e}`);
      }
      throw e;
    }
  }

  async headObject({
    context,
    key,
    schemaVersionId,
  }: {
    context: Context;
    key: string;
    schemaVersionId: string;
  }): Promise<boolean> {
    const command = new HeadObjectCommand({
      Bucket: this.bucketName,
      Key: key,
    });

    try {
      const resp = await this.s3Client.send(command);
      if (resp.$metadata.httpStatusCode === 404) {
        throw new BlobNotFoundError(`Object not found`);
      } else if (resp.$metadata.httpStatusCode === 304) {
        return false;
      } else if (resp.$metadata.httpStatusCode !== 200) {
        throw new Error(`Failed to fetch the metadata of the object.`);
      }
      if (resp.Metadata && resp.Metadata.version === schemaVersionId) {
        return false;
      }
      return true;
    } catch (e: any) {
      if (e instanceof NoSuchKey || e instanceof NotFound) {
        throw new BlobNotFoundError(`Object not found: ${e}`);
      }
      throw e;
    }
  }
}

export const createS3BlobStorage = (storageUrl: string): BlobStorage => {
  const url = new URL(storageUrl);
  const region = url.searchParams.get('region') ?? 'default';
  const s3Client = new S3Client({
    region,
    endpoint: url.origin,
    credentials: {
      accessKeyId: url.username ?? '',
      secretAccessKey: url.password ?? '',
    },
    forcePathStyle: true,
  });
  const bucketName = url.pathname.slice(1);
  return new S3BlobStorage(s3Client, bucketName);
};
