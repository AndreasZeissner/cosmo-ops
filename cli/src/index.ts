#!/usr/bin/env node

import program from './commands/index.js';

await program.parseAsync(process.argv);
