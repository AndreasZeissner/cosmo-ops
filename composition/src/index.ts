export * from './ast/ast';
export * from './ast/utils';
export * from './errors/errors';
export * from './federation/federation-factory';
export * from './federation/utils';
export * from './normalization/normalization-factory';
export * from './router-configuration/router-configuration';
export * from './subgraph/subgraph';
export * from './schema-building/ast';
export * from './schema-building/type-definition-data';
export * from './schema-building/type-merging';
export * from './utils/utils';