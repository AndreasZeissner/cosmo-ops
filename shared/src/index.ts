export * from './utils/util.js';
export * from './router-config/builder.js';
export * from './openai-graphql/index.js';
export * from './events/types.js';
