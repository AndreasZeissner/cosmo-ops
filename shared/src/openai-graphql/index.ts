import { Configuration, OpenAIApi } from 'openai';
import { z } from 'zod';
import { zodToJsonSchema } from 'zod-to-json-schema';

export class OpenAIGraphql {
  private readonly client: OpenAIApi;

  constructor(config: { openAiApiKey: string | undefined }) {
    const configuration = new Configuration({
      apiKey: config.openAiApiKey,
    });
    this.client = new OpenAIApi(configuration);
  }

  public async fixSDL(input: { sdl: string; checkResult: string }): Promise<{ sdl: string }> {
    const out = z.object({
      sdl: z.string().describe('The fixed GraphQL Schema'),
    });
    const res = await this.client.createChatCompletion({
      model: 'gpt-3.5-turbo',
      temperature: 0,
      messages: [
        {
          role: 'user',
          content: `Given the following SDL (Schema Definition Language) from an Apollo Federation subgraph:
                    
                    \`\`\`graphql
                    ${input.sdl}
                    \`\`\`
                    
                    I've got the following schema composition result back from rover:
                    
                    \`\`\`
                    ${input.checkResult}
                    \`\`\`
                    
                    Please take into consideration the schema check result and generate a fixed version of the SDL.
                    Once completed, store the SDL using the 'store_sdl' function.`,
        },
      ],
      functions: [
        {
          name: 'store_sdl',
          description:
            'This function allows the Agent to store the fixed SDL. The SDL must be returned in plain text as a GraphQL Schema.',
          parameters: zodToJsonSchema(out),
        },
      ],
    });

    try {
      return out.parse(JSON.parse(res.data.choices[0].message!.function_call!.arguments!));
    } catch (e: any) {
      const errorText = e.toString();
      if (errorText.includes('Unexpected token')) {
        return this.fixSDL(input);
      }
      console.log(`OpenAI fixSDL failed: ${errorText}\n\nresponse:\n\n${JSON.stringify(res.data)}`);
      throw new Error('OpenAI fixSDL failed');
    }
  }
}
