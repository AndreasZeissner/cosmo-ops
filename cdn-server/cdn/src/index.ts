import { JWTVerifyResult, jwtVerify } from 'jose';
import { Context, Env, Hono, Next, Schema } from 'hono';
import { streamToJSON } from './utils';

export interface BlobStorage {
  getObject({
    context,
    key,
    cacheControl,
  }: {
    context: Context;
    key: string;
    cacheControl?: string;
  }): Promise<ReadableStream>;
}

export class BlobNotFoundError extends Error {
  constructor(message: string, cause?: Error) {
    super(message, cause);
    Object.setPrototypeOf(this, BlobNotFoundError.prototype);
  }
}

interface CdnOptions {
  authJwtSecret: string | ((c: Context) => string);
  blobStorage: BlobStorage;
}

declare module 'hono' {
  interface ContextVariableMap {
    authenticatedOrganizationId: string;
    authenticatedFederatedGraphId: string;
  }
}

const jwtMiddleware = (secret: string | ((c: Context) => string)) => {
  return async (c: Context, next: Next) => {
    const authHeader = c.req.header('Authorization');
    if (!authHeader) {
      return c.text('Unauthorized', 401);
    }
    const [type, token] = authHeader.split(' ');
    if (type !== 'Bearer' || !token) {
      return c.text('Unauthorized', 401);
    }
    let result: JWTVerifyResult;
    const secretKey = new TextEncoder().encode(typeof secret === 'function' ? secret(c) : secret);
    try {
      result = await jwtVerify(token, secretKey);
    } catch (e: any) {
      if (e instanceof Error && (e.name === 'JWSSignatureVerificationFailed' || e.name === 'JWSInvalid')) {
        return c.text('Unauthorized', 401);
      }
      throw e;
    }
    const organizationId = result.payload.organization_id;
    const federatedGraphId = result.payload.federated_graph_id;
    if (!organizationId || !federatedGraphId) {
      return c.text('Forbidden', 403);
    }
    c.set('authenticatedOrganizationId', organizationId);
    c.set('authenticatedFederatedGraphId', federatedGraphId);
    await next();
  };
};

const persistedOperation = (storage: BlobStorage) => {
  return async (c: Context) => {
    const organizationId = c.req.param('organization_id');
    const federatedGraphId = c.req.param('federated_graph_id');
    // Check authentication
    if (
      organizationId !== c.get('authenticatedOrganizationId') ||
      federatedGraphId !== c.get('authenticatedFederatedGraphId')
    ) {
      return c.text('Bad Request', 400);
    }
    const clientId = c.req.param('client_id');
    const operation = c.req.param('operation');
    if (!operation.endsWith('.json')) {
      return c.notFound();
    }
    const key = `${organizationId}/${federatedGraphId}/operations/${clientId}/${operation}`;
    let operationStream: ReadableStream;
    try {
      operationStream = await storage.getObject({ context: c, key });
    } catch (e: any) {
      if (e instanceof Error && e.constructor.name === 'BlobNotFoundError') {
        return c.notFound();
      }
      throw e;
    }
    return c.stream(async (stream) => {
      await stream.pipe(operationStream);
      await stream.close();
    });
  };
};

const routerConfig = (storage: BlobStorage) => {
  return async (c: Context) => {
    const organizationId = c.req.param('organization_id');
    const federatedGraphId = c.req.param('federated_graph_id');
    // Check authentication
    if (
      organizationId !== c.get('authenticatedOrganizationId') ||
      federatedGraphId !== c.get('authenticatedFederatedGraphId')
    ) {
      return c.text('Bad Request', 400);
    }
    const key = `${organizationId}/${federatedGraphId}/routerconfigs/latest.json`;
    let configStream: ReadableStream;
    try {
      configStream = await storage.getObject({ context: c, key, cacheControl: 'no-cache' });
    } catch (e: any) {
      if (e instanceof Error && e.constructor.name === 'BlobNotFoundError') {
        return c.notFound();
      }
      throw e;
    }

    // Because we work with a fixed path of /routerconfigs/latest.json
    // we need to enforce no-cache headers here.
    c.header('Cache-Control', 'private, no-cache, no-store, max-age=0, must-revalidate');
    c.header('Content-Type', 'application/json; charset=utf-8');

    const teedStream = configStream.tee();

    const routerConfig = await streamToJSON(teedStream[0]);
    const body = await c.req.json();

    if (body?.version === undefined || null) {
      return c.text('Bad Request', 400);
    }

    if (routerConfig?.version && body.version === routerConfig.version) {
      return c.body(null, 204);
    }

    return c.stream(async (stream) => {
      await stream.pipe(teedStream[1]);
      await stream.close();
    });
  };
};

// eslint-disable-next-line @typescript-eslint/ban-types
export const cdn = <E extends Env, S extends Schema = {}, BasePath extends string = '/'>(
  hono: Hono<E, S, BasePath>,
  opts: CdnOptions,
) => {
  const operations = '/:organization_id/:federated_graph_id/operations/:client_id/:operation{.+\\.json$}';
  hono.use(operations, jwtMiddleware(opts.authJwtSecret));
  hono.get(operations, persistedOperation(opts.blobStorage));

  const routerConfigs = '/:organization_id/:federated_graph_id/routerconfigs/latest.json';
  hono.use(routerConfigs, jwtMiddleware(opts.authJwtSecret));
  hono.post(routerConfigs, routerConfig(opts.blobStorage));
};
