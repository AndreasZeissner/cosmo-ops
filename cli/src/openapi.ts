import type { OpenAPIV3, OpenAPIV2 } from 'openapi-types';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { getJSONSchemaOptionsFromOpenAPIOptions } from '@omnigraph/openapi';
import { loadNonExecutableGraphQLSchemaFromJSONSchemas } from '@omnigraph/json-schema';

type OasSource = string | OpenAPIV3.Document | OpenAPIV2.Document;

export type IntrospectOpenApiOptions = {
  source: OasSource;
  name: string;
  cwd?: string;
};

export const introspectOpenApi = async (options: IntrospectOpenApiOptions): Promise<string> => {
  const { cwd = process.cwd(), ...rest } = options;

  // get json schema options describing each path in the spec
  const extraJSONSchemaOptions = await getJSONSchemaOptionsFromOpenAPIOptions(options.name, options);
  // build graphql schema from the json schema options with the directives attached to fields
  const graphQLSchema = await loadNonExecutableGraphQLSchemaFromJSONSchemas(options.name, {
    ...rest,
    ...extraJSONSchemaOptions,
    cwd,
  });

  // as logic of translating api calls stored in the directives we need print schema with directives
  return printSchemaWithDirectives(graphQLSchema);
};
