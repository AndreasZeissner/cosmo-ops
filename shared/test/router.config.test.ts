import * as fs from 'node:fs';
import * as path from 'node:path';
import * as url from 'node:url';
import { describe, expect, test } from 'vitest';
import { buildRouterConfig, Subgraph } from '../src';

// @ts-ignore-next-line
const __dirname = url.fileURLToPath(new URL(".", import.meta.url));

describe("Router Config Builder", () => {
  test("Build Subgraph schema", () => {
    const accounts: Subgraph = {
      sdl: fs.readFileSync(
        path.join(__dirname, "testdata", "accounts.graphql"),
        {
          encoding: "utf8",
        }
      ),
      url: "https://wg-federation-demo-accounts.fly.dev/graphql",
    };
    const products: Subgraph = {
      sdl: fs.readFileSync(
        path.join(__dirname, "testdata", "products.graphql"),
        {
          encoding: "utf8",
        }
      ),
      url: "https://wg-federation-demo-products.fly.dev/graphql",
    };
    const reviews: Subgraph = {
      sdl: fs.readFileSync(
        path.join(__dirname, "testdata", "reviews.graphql"),
        {
          encoding: "utf8",
        }
      ),
      url: "https://wg-federation-demo-reviews.fly.dev/graphql",
    };
    const inventory: Subgraph = {
      sdl: fs.readFileSync(
        path.join(__dirname, "testdata", "inventory.graphql"),
        {
          encoding: "utf8",
        }
      ),
      url: "https://wg-federation-demo-inventory.fly.dev/graphql",
    };
    const routerConfig = buildRouterConfig({
      subgraphs: [accounts, products, reviews, inventory],
      // Passed as it is to the router config
      federatedSDL: `type Query {}`,
    });
    const json = routerConfig.toJsonString({
      enumAsInteger: true,
      emitDefaultValues: false,
    });
    const out = JSON.stringify(JSON.parse(json), null, 2);
    expect(out).matchSnapshot("router.config.json");
  });

  test('that builder config throws an error if the graph fails normalization', () => {
    const subgraph = {
      sdl: `extend input Human {
        name: String!
      }`,
      url: '',
    };
    let error;
    try {
      buildRouterConfig({
        subgraphs: [subgraph],
        federatedSDL: '',
      });
    } catch (e) {
      error = e;
    }
    expect(error).toBeDefined();
    expect(error.message).toBe('Normalization failed')
    expect(error.cause.message).toBe('Extension error:\n' +
      ' Could not extend the type "Human" because no base definition exists.');
  });
});
