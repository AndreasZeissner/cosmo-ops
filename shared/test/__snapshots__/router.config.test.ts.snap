// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Router Config Builder > Build Subgraph schema > router.config.json 1`] = `
"{
  \\"engineConfig\\": {
    \\"defaultFlushInterval\\": \\"500\\",
    \\"datasourceConfigurations\\": [
      {
        \\"kind\\": 1,
        \\"rootNodes\\": [
          {
            \\"typeName\\": \\"Query\\",
            \\"fieldNames\\": [
              \\"me\\"
            ]
          },
          {
            \\"typeName\\": \\"User\\",
            \\"fieldNames\\": [
              \\"id\\",
              \\"name\\",
              \\"username\\"
            ]
          }
        ],
        \\"overrideFieldPathFromAlias\\": true,
        \\"customGraphql\\": {
          \\"fetch\\": {
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-accounts.fly.dev/graphql\\"
            },
            \\"method\\": 1,
            \\"body\\": {},
            \\"baseUrl\\": {},
            \\"path\\": {}
          },
          \\"subscription\\": {
            \\"enabled\\": true,
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-accounts.fly.dev/graphql\\"
            }
          },
          \\"federation\\": {
            \\"enabled\\": true,
            \\"serviceSdl\\": \\"type Query @extends {\\\\n    me: User\\\\n}\\\\n\\\\ntype User @key(fields: \\\\\\"id\\\\\\") {\\\\n    id: ID!\\\\n    name: String\\\\n    username: String\\\\n}\\\\n\\"
          },
          \\"upstreamSchema\\": {
            \\"key\\": \\"56b22d2b228499be9dcb1bd0720bbf7d8348104e\\"
          }
        },
        \\"requestTimeoutSeconds\\": \\"10\\",
        \\"id\\": \\"https://wg-federation-demo-accounts.fly.dev/graphql\\",
        \\"keys\\": [
          {
            \\"typeName\\": \\"User\\",
            \\"selectionSet\\": \\"id\\"
          }
        ]
      },
      {
        \\"kind\\": 1,
        \\"rootNodes\\": [
          {
            \\"typeName\\": \\"Query\\",
            \\"fieldNames\\": [
              \\"topProducts\\"
            ]
          },
          {
            \\"typeName\\": \\"Subscription\\",
            \\"fieldNames\\": [
              \\"updatedPrice\\"
            ]
          },
          {
            \\"typeName\\": \\"Mutation\\",
            \\"fieldNames\\": [
              \\"setPrice\\"
            ]
          },
          {
            \\"typeName\\": \\"Product\\",
            \\"fieldNames\\": [
              \\"upc\\",
              \\"name\\",
              \\"price\\",
              \\"weight\\"
            ]
          }
        ],
        \\"overrideFieldPathFromAlias\\": true,
        \\"customGraphql\\": {
          \\"fetch\\": {
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-products.fly.dev/graphql\\"
            },
            \\"method\\": 1,
            \\"body\\": {},
            \\"baseUrl\\": {},
            \\"path\\": {}
          },
          \\"subscription\\": {
            \\"enabled\\": true,
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-products.fly.dev/graphql\\"
            }
          },
          \\"federation\\": {
            \\"enabled\\": true,
            \\"serviceSdl\\": \\"type Query @extends {\\\\n    topProducts(first: Int = 5, random: Boolean): [Product]\\\\n}\\\\n\\\\ntype Product @key(fields: \\\\\\"upc\\\\\\") {\\\\n    upc: String!\\\\n    name: String\\\\n    price: Int\\\\n    weight: Int\\\\n}\\\\n\\\\ntype Subscription @extends {\\\\n    updatedPrice: Product!\\\\n}\\\\n\\\\ntype Mutation @extends {\\\\n    setPrice(upc: String!, price: Int!): Product\\\\n}\\\\n\\"
          },
          \\"upstreamSchema\\": {
            \\"key\\": \\"81ea6c308ed2bc651346d855b52a0bfa15303236\\"
          }
        },
        \\"requestTimeoutSeconds\\": \\"10\\",
        \\"id\\": \\"https://wg-federation-demo-products.fly.dev/graphql\\",
        \\"keys\\": [
          {
            \\"typeName\\": \\"Product\\",
            \\"selectionSet\\": \\"upc\\"
          }
        ]
      },
      {
        \\"kind\\": 1,
        \\"rootNodes\\": [
          {
            \\"typeName\\": \\"User\\",
            \\"fieldNames\\": [
              \\"id\\",
              \\"username\\",
              \\"reviews\\"
            ]
          },
          {
            \\"typeName\\": \\"Product\\",
            \\"fieldNames\\": [
              \\"upc\\",
              \\"reviews\\"
            ]
          },
          {
            \\"typeName\\": \\"Review\\",
            \\"fieldNames\\": [
              \\"id\\",
              \\"body\\",
              \\"author\\",
              \\"product\\"
            ]
          }
        ],
        \\"overrideFieldPathFromAlias\\": true,
        \\"customGraphql\\": {
          \\"fetch\\": {
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-reviews.fly.dev/graphql\\"
            },
            \\"method\\": 1,
            \\"body\\": {},
            \\"baseUrl\\": {},
            \\"path\\": {}
          },
          \\"subscription\\": {
            \\"enabled\\": true,
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-reviews.fly.dev/graphql\\"
            }
          },
          \\"federation\\": {
            \\"enabled\\": true,
            \\"serviceSdl\\": \\"type Review @key(fields: \\\\\\"id\\\\\\") {\\\\n    id: ID!\\\\n    body: String\\\\n    author: User @provides(fields: \\\\\\"username\\\\\\")\\\\n    product: Product\\\\n}\\\\n\\\\ntype User @extends @key(fields: \\\\\\"id\\\\\\") {\\\\n    id: ID! @external\\\\n    username: String @external\\\\n    reviews: [Review]\\\\n}\\\\n\\\\ntype Product @extends @key(fields: \\\\\\"upc\\\\\\") {\\\\n    upc: String! @external\\\\n    reviews: [Review]\\\\n}\\\\n\\"
          },
          \\"upstreamSchema\\": {
            \\"key\\": \\"14585f5697286d11d971ee9c4dc02426caa838d5\\"
          }
        },
        \\"requestTimeoutSeconds\\": \\"10\\",
        \\"id\\": \\"https://wg-federation-demo-reviews.fly.dev/graphql\\",
        \\"keys\\": [
          {
            \\"typeName\\": \\"User\\",
            \\"selectionSet\\": \\"id\\"
          },
          {
            \\"typeName\\": \\"Product\\",
            \\"selectionSet\\": \\"upc\\"
          },
          {
            \\"typeName\\": \\"Review\\",
            \\"selectionSet\\": \\"id\\"
          }
        ]
      },
      {
        \\"kind\\": 1,
        \\"rootNodes\\": [
          {
            \\"typeName\\": \\"Product\\",
            \\"fieldNames\\": [
              \\"upc\\",
              \\"weight\\",
              \\"price\\",
              \\"inStock\\",
              \\"shippingEstimate\\"
            ]
          }
        ],
        \\"overrideFieldPathFromAlias\\": true,
        \\"customGraphql\\": {
          \\"fetch\\": {
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-inventory.fly.dev/graphql\\"
            },
            \\"method\\": 1,
            \\"body\\": {},
            \\"baseUrl\\": {},
            \\"path\\": {}
          },
          \\"subscription\\": {
            \\"enabled\\": true,
            \\"url\\": {
              \\"staticVariableContent\\": \\"https://wg-federation-demo-inventory.fly.dev/graphql\\"
            }
          },
          \\"federation\\": {
            \\"enabled\\": true,
            \\"serviceSdl\\": \\"type Product @extends @key(fields: \\\\\\"upc\\\\\\") {\\\\n    upc: String! @external\\\\n    weight: Int @external\\\\n    price: Int @external\\\\n    inStock: Boolean\\\\n    shippingEstimate: Int @requires(fields: \\\\\\"price weight\\\\\\")\\\\n}\\\\n\\"
          },
          \\"upstreamSchema\\": {
            \\"key\\": \\"48302ab241606707e48c307cf5ec547c0a2780fe\\"
          }
        },
        \\"requestTimeoutSeconds\\": \\"10\\",
        \\"id\\": \\"https://wg-federation-demo-inventory.fly.dev/graphql\\",
        \\"keys\\": [
          {
            \\"typeName\\": \\"Product\\",
            \\"selectionSet\\": \\"upc\\"
          }
        ]
      }
    ],
    \\"graphqlSchema\\": \\"type Query {}\\",
    \\"stringStorage\\": {
      \\"56b22d2b228499be9dcb1bd0720bbf7d8348104e\\": \\"schema {\\\\n  query: Query\\\\n}\\\\n\\\\ndirective @extends on INTERFACE | OBJECT\\\\n\\\\ndirective @external on FIELD_DEFINITION | OBJECT\\\\n\\\\ndirective @key(fields: String!) repeatable on OBJECT\\\\n\\\\ndirective @provides(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @requires(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @tag(name: String!) repeatable on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION\\\\n\\\\ntype Query {\\\\n  me: User\\\\n}\\\\n\\\\ntype User @key(fields: \\\\\\"id\\\\\\") {\\\\n  id: ID!\\\\n  name: String\\\\n  username: String\\\\n}\\",
      \\"81ea6c308ed2bc651346d855b52a0bfa15303236\\": \\"schema {\\\\n  query: Query\\\\n  mutation: Mutation\\\\n  subscription: Subscription\\\\n}\\\\n\\\\ndirective @extends on INTERFACE | OBJECT\\\\n\\\\ndirective @external on FIELD_DEFINITION | OBJECT\\\\n\\\\ndirective @key(fields: String!) repeatable on OBJECT\\\\n\\\\ndirective @provides(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @requires(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @tag(name: String!) repeatable on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION\\\\n\\\\ntype Mutation {\\\\n  setPrice(price: Int!, upc: String!): Product\\\\n}\\\\n\\\\ntype Product @key(fields: \\\\\\"upc\\\\\\") {\\\\n  name: String\\\\n  price: Int\\\\n  upc: String!\\\\n  weight: Int\\\\n}\\\\n\\\\ntype Query {\\\\n  topProducts(first: Int = 5, random: Boolean): [Product]\\\\n}\\\\n\\\\ntype Subscription {\\\\n  updatedPrice: Product!\\\\n}\\",
      \\"14585f5697286d11d971ee9c4dc02426caa838d5\\": \\"directive @extends on INTERFACE | OBJECT\\\\n\\\\ndirective @external on FIELD_DEFINITION | OBJECT\\\\n\\\\ndirective @key(fields: String!) repeatable on OBJECT\\\\n\\\\ndirective @provides(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @requires(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @tag(name: String!) repeatable on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION\\\\n\\\\ntype Product @key(fields: \\\\\\"upc\\\\\\") {\\\\n  reviews: [Review]\\\\n  upc: String! @external\\\\n}\\\\n\\\\ntype Review @key(fields: \\\\\\"id\\\\\\") {\\\\n  author: User @provides(fields: \\\\\\"username\\\\\\")\\\\n  body: String\\\\n  id: ID!\\\\n  product: Product\\\\n}\\\\n\\\\ntype User @key(fields: \\\\\\"id\\\\\\") {\\\\n  id: ID! @external\\\\n  reviews: [Review]\\\\n  username: String @external\\\\n}\\",
      \\"48302ab241606707e48c307cf5ec547c0a2780fe\\": \\"directive @extends on INTERFACE | OBJECT\\\\n\\\\ndirective @external on FIELD_DEFINITION | OBJECT\\\\n\\\\ndirective @key(fields: String!) repeatable on OBJECT\\\\n\\\\ndirective @provides(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @requires(fields: String!) on FIELD_DEFINITION\\\\n\\\\ndirective @tag(name: String!) repeatable on ARGUMENT_DEFINITION | ENUM | ENUM_VALUE | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | INPUT_OBJECT | INTERFACE | OBJECT | SCALAR | UNION\\\\n\\\\ntype Product @key(fields: \\\\\\"upc\\\\\\") {\\\\n  inStock: Boolean\\\\n  price: Int @external\\\\n  shippingEstimate: Int @requires(fields: \\\\\\"price weight\\\\\\")\\\\n  upc: String! @external\\\\n  weight: Int @external\\\\n}\\"
    }
  }
}"
`;
